core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/function/FunctionDSL.java

```diff
@@ -0,0 +1,160 @@
+import com.amazon.opendistroforelasticsearch.sql.expression.FunctionExpression;
+import com.amazon.opendistroforelasticsearch.sql.expression.env.Environment;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import lombok.experimental.UtilityClass;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * Function Define Utility.
+ */
+@UtilityClass
+public class FunctionDSL {
+  /**
+   * Define overloaded function with implementation.
+   *
+   * @param functionName function name.
+   * @param functions    a list of function implementation.
+   * @return FunctionResolver.
+   */
+  public FunctionResolver define(FunctionName functionName,
+                                 Function<FunctionName, Pair<FunctionSignature,
+                                     FunctionBuilder>>... functions) {
+    return define(functionName, Arrays.asList(functions));
+  }
+
+  /**
+   * Define overloaded function with implementation.
+   *
+   * @param functionName function name.
+   * @param functions    a list of function implementation.
+   * @return FunctionResolver.
+   */
+  public FunctionResolver define(FunctionName functionName,
+                                 List<Function<FunctionName, Pair<FunctionSignature,
+                                     FunctionBuilder>>> functions) {
+
+    FunctionResolver.FunctionResolverBuilder builder = FunctionResolver.builder();
+    builder.functionName(functionName);
+    for (Function<FunctionName, Pair<FunctionSignature, FunctionBuilder>> func : functions) {
+      Pair<FunctionSignature, FunctionBuilder> functionBuilder = func.apply(functionName);
+      builder.functionBundle(functionBuilder.getKey(), functionBuilder.getValue());
+    }
+    return builder.build();
+  }
+
+  /**
+   * Unary Function Implementation.
+   *
+   * @param function   {@link ExprValue} based unary function.
+   * @param returnType return type.
+   * @param argsType   argument type.
+   * @return Unary Function Implementation.
+   */
+  public SerializableFunction<FunctionName, Pair<FunctionSignature, FunctionBuilder>> impl(
+      SerializableFunction<ExprValue, ExprValue> function,
+      ExprType returnType,
+      ExprType argsType) {
+
+    return functionName -> {
+      FunctionSignature functionSignature =
+          new FunctionSignature(functionName, Collections.singletonList(argsType));
+      FunctionBuilder functionBuilder =
+          arguments -> new FunctionExpression(functionName, arguments) {
+            @Override
+            public ExprValue valueOf(Environment<Expression, ExprValue> valueEnv) {
+              ExprValue value = arguments.get(0).valueOf(valueEnv);
+              if (value.isMissing()) {
+                return ExprValueUtils.missingValue();
+              } else if (value.isNull()) {
+                return ExprValueUtils.nullValue();
```

Comment:
I may miss it somewhere. Where is this MISSING/NULL check logic for the `impl()` below?

Reply:
Good points.

Remove the NULL/MISSING generic handling logic in FunctionDSL.impl, and provide another utilit method nullMissingHandling to the client. The reason for doing that is (1) most of operator/function follow the generic missing/null handling logic. (2) but some doesn't follow that e.g. and/or/xor. 
```
  public SerializableFunction<ExprValue, ExprValue> nullMissingHandling(
      SerializableFunction<ExprValue, ExprValue> function) {
    return value -> {
      if (value.isMissing()) {
        return ExprValueUtils.missingValue();
      } else if (value.isNull()) {
        return ExprValueUtils.nullValue();
      } else {
        return function.apply(value);
      }
    };
  }
```
The client could make decision whetehr it want to follow the generic rule or not.
```
FunctionDSL.impl(FunctionDSL.nullMissingHandling(DateTimeFunction::exprDayOfMonth),
            INTEGER, DATE)
```